documentation/modules/configuring/con-config-mirrormaker2-high-volume-messages.adoc

```diff
@@ -6,25 +6,24 @@
 = Handling high volumes of messages
 
 [role="_abstract"]
-If your MirrorMaker 2.0 deployment is going to be handling a high volume of messages, you might need to adjust its configuration to support it. The configuration can help manage throughput and issues with latency. 
+If your MirrorMaker 2.0 deployment is going to be handling a high volume of messages, you might need to adjust its configuration to support it. 
+The configuration can help manage throughput and issues with latency. 
 
-MirrorMaker 2.0 uses an internal producer client (`KafkaProducer`) to replicate data from the source to target Kafka cluster. 
+MirrorMaker 2.0 uses an internal producer client to replicate data from the source to the target Kafka cluster. 
 To control the streaming of high volumes of data, you can tune the following producer configuration for optimal throughput: 
 
 * Producer batch size 
-* Producer buffer size 
+* Producer buffer memory 
 
 .Producer batch size
 In certain circumstances, you might want to increase the size of the message batches sent in a single produce request.
 You increase the batch size using the `batch.size` producer configuration to reduce the number of outstanding messages ready to be sent.
 A larger batch size reduces the number of requests and the backlog in the message queue.
 A produce request is sent to the target cluster when the batch size is reached.
 By increasing the batch size, produce requests are delayed and more messages are added to the batch and sent to brokers at the same time.  
-This can improve throughput when you have just a few topic partitions that handle large numbers of messages. 
+This can improve throughput when you have just a few topic partitions that handle large numbers of messages.  
 
-NOTE: To be able to increase the batch size, the `connector.client.config.override.policy` must be set to `All`. This is the default setting. The property enables connector client configurations.   
-
-.Producer buffer size
+.Producer buffer memory
```

Comment:
I thought we just wanted to provide some tuning advice here, but if we want to include the root cause analysis, then it would be something like this, with the note about recent changes in that space:

When there is a high volume of messages, mirroring may be slow or not working at all.
This happens because the underline **Producer is not optimized for throughput**.

At every iteration, the MirrorMaker2 flush pipeline is the following:

```
source-topic -> Connect outstanding messages queue (# records) -> producer buffer (bytes) -> target-topic
```

The `MirrorSourceTask` periodically `poll()` from the source topic, even before the previous batch of records has been acknowledged.
In the meantime, a different thread tries to commit the source topic offsets every `offset.flush.timeout.ms`.
One key point here is that the commit process waits until the outstanding messages queue is empty to proceed.

When there is a high volume of messages coming in, the outstanding messages queue can grow unbounded (there is no hard limit).
This causes the Producer's buffer to fill up and the `send()` operation to become synchronous, which also prevents the `MirrorSourceTask` from polling.
Eventually, the Producer becomes unable to free that queue before the commit times out and fails.

>Since Kafka 3.1.0 (Strimzi 0.28.0), the offsets commit does not wait for the producer to empty the queue and the `offset.flush.timeout.ms` only applies to the actual writing of offset data to the internal offsets topic.
> Mirroring can still be very slow when having a high volume, so the optimization proposed here still applies.

Reply:
Thanks @federico. I think we cover the config that handles the pipeline in enough detail for context, but please add any specific suggestions if you think we are missing something.

Reply:
There is still no mention of recent changes (see the quote at the end of my previous message). We should also remove `producer.buffer.memory: 8388608` from the example.

Reply:
Thanks @fvaleri  -- I added a note and removed `producer.buffer.memory` from the example

Reply:
Also added linger

Reply:
Great, it looks almost good.

For `producer.linger.ms` I was searching for a confirmation, but probably it is not directly related to the problem at hand, so I would leave it out.